### ISSUE TITLE
Add Ftp(s)-trigger

### ISSUE THREAD

**Issue by iremmats** [Azure/azure-webjobs-sdk-extensions#2]
Just like FileTrigger, add one for ftp and ftps.

**Comment by mathewc** [comment]
Can you please elaborate on the features/functionality you envision for such a binding? E.g., poling model, attributes + their knobs/settings, etc.

There are lots of possibilities for new triggers/binders, and we're expecting the community will start writing the extensions they want for the benefit of all :) Are you thinking of taking a stab at this one? ;)

In this particular case, it might be that an FTP extension belongs in the Files extension already in this repo - i.e., the new bindings for FTP can just be added to the Files extension, and leverage the existing config.UseFiles() registration, etc.

**Comment by iremmats** [comment]
Are you guys working on this one yet? Otherwise Ill start working on this one next week.

**Comment by mathewc** [comment]
No, we're not working on this - it's fair game! Once you get to the point where you have the programming model ironed out (attributes, etc.) can you please share for review/feedback? Thanks :)

**Comment by iremmats** [comment]
Of course, just let me know the procedure for review/feedback! :)
______________________________
Mats Iremark, Cloud Architect @ e-man AB
Tel: 0708-194243

Från: Mathew Charles <notifications@github.com<mailto:notifications@github.com>>
Svara till: Azure/azure-webjobs-sdk-extensions <reply@reply.github.com<mailto:reply@reply.github.com>>
Datum: Friday 28 August 2015 18:25
Till: Azure/azure-webjobs-sdk-extensions <azure-webjobs-sdk-extensions@noreply.github.com<mailto:azure-webjobs-sdk-extensions@noreply.github.com>>
Kopia: Mats Iremark <mats.iremark@e-man.se<mailto:mats.iremark@e-man.se>>
Ämne: Re: [azure-webjobs-sdk-extensions] Add Ftp(s)-trigger (#2)


No, we're not working on this - it's fair game! Once you get to the point where you have the programming model ironed out (attributes, etc.) can you please share for review/feedback? Thanks :)

-

**Comment by mathewc** [comment]
I wasn't thinking anything heavyweight - perhaps just here on this issue, or via code in your fork.

**Comment by iremmats** [comment]
Ok Ill post here and keep you updated.
______________________________
Mats Iremark, Cloud Architect @ e-man AB
Tel: 0708-194243

Från: Mathew Charles <notifications@github.com<mailto:notifications@github.com>>
Svara till: Azure/azure-webjobs-sdk-extensions <reply@reply.github.com<mailto:reply@reply.github.com>>
Datum: Friday 28 August 2015 18:34
Till: Azure/azure-webjobs-sdk-extensions <azure-webjobs-sdk-extensions@noreply.github.com<mailto:azure-webjobs-sdk-extensions@noreply.github.com>>
Kopia: Mats Iremark <mats.iremark@e-man.se<mailto:mats.iremark@e-man.se>>
Ämne: Re: [azure-webjobs-sdk-extensions] Add Ftp(s)-trigger (#2)


I wasn't thinking anything heavyweight - perhaps just here on this issue, or via code in your fork.

-

**Comment by iremmats** [comment]
I threw up a repo and followed the guide how to create an extension. Still a little bit difficult trying to understand how its all hanging together... :)
So far its an ftp trigger only. Ill try to add the non-trigger too.
https://github.com/iremmats/webjobs-ftp-extension

The console app has one function that triggers on files in a directory on a server. 

Filemask, dont delete files but move them to a "processed" directory and other stuff is planned. Error handling hasn't been touched yet...
Let me know what you think...

**Comment by mathewc** [comment]
Nice start! Here is some initial quick feedback.

* Can all the connection info server/uid/pwd be defaulted from config (FtpConfiguration) so it doesn't have to be specified each time? That would make sense if the common pattern will be to interact with a single server. Might allow the ability for attribute to override the connection info.
* Can you use a standard ftp connection format like ftp://username:password@domain.com so there is only a single connection string that needs to be specified by the user (rather than 3 separate strings)?
* FtpTriggerValue shouldn't be public. The user will instead bind to Stream/FileStream/string, etc. as in the FileTrigger
* Path/FileMask might not belong on FtpConfiguration - those seem like per function settings
* FtpTriggerAttribute should have a ctor that takes the required bits (e.g. Path)
* The ArxOne FTP library you're using - is it the most popular/stable? How did you choose it?
* I assume you also plan on an Ftp binding that reads/writes single files (non-trigger binding)?
* Still need to finish Dashboard integration of course :)
* Is there an FtpListener class/library that you can use to do the listening?
* You need to think about multi-instance. The listener as is, when scaled out, will mean multiple jobs will be processing the same file. Handling this right is complicated - you can see the status file approach I took with FileTrigger.
* In your custom polling logic, you currently pull ALL files and delete them when finished. People will probably also want to be able to have their files processed, but not deleted.
* Will probably want to make the polling interval customizable on the global config
* You'll need to make the invocation call from the listener to the function async, otherwise you won't get any parallelism. See how I used an async work queue for this in FileTrigger - that also gives the benefit of being able to manage work item concurrency.

**Comment by iremmats** [comment]
Thanks! I will try to implement those. Will keep you updated.

**Comment by iremmats** [comment]
I took some time and made it more like the FileTrigger. Its in not even close to runable state now though.

* Ill follow your advice on the configuration parts. Im starting to figure out how everything is connected.
Default from FtpConfiguration. One connection string just AzureWebJobsServiceBus etc.

* Regarding the FtpTriggerValue, I need not just the content of the file but also its name. Many scenarios we are working with, file name is important. Please help me here with advice on how to do it. Ill bind it to string, stream and maybe FileInfo?

* Path/Filemask will be moved to each function. Ill check the FileTrigger code for guidelines.

* Ctor for FtpTriggerAttribute added.
* ArxOneFtp, just googled it. Ill probably replace it. You prefer I write from scratch?
* Dashboard integration is in the backlog.
* I found a FtpListener example online and put it into the repo. Check if can be used.
* Regarding multiinstance Ill take your StatusFile approach. Just have to figure out if it needs any changes.
* Most of the time "get and delete" is the preferred scenario. Another solution we often use is moving to a processed folder.
*Polling interval customizable, check. If I get time Ill add the same kind of cron timer as you have for scheduling jobs.
*Will make it async.

**Comment by mack0196** [comment]
Any progress on the Azure WebJob FTP trigger. I am looking for a solution to trigger a webjob when a new FTP file is done writing.

**Comment by StefH** [comment]
In case anyone is still looking for a solution, I created a new project to implement an FTP trigger.

See https://github.com/StefH/WebJobs.Extensions.Ftp which supports:
- Ftp Trigger on a new FtpFile, FtpFile[], FtpStream and FtpStream[]
- Ftp Bindings on: IAsyncCollector, IFtpClient, FtpFile and FtpStream

@iremmats : if you still have some time, please review my code
